src/xrpld/app/tx/detail/NFTokenUtils.cpp

```diff
@@ -238,6 +238,39 @@ compareTokens(uint256 const& a, uint256 const& b)
     return a < b;
 }
 
+TER
+changeTokenURI(
+    ApplyView& view,
+    AccountID const& owner,
+    uint256 const& nftokenID,
+    std::optional<ripple::Slice> const& uri)
```

Comment:
@scottschurr may I ask why is `Slice` type used here, and not `Buffer`?

Also another related question, in `STBlob`, it uses`Buffer` to store `value_`, but why `value_type` have type `Slice` and not `Buffer`? In `STBlob`:
```
class STBlob : public STBase, public CountedObject<STBlob>
{
    Buffer value_;

public:
    using value_type = Slice;
```

Thanks!

Reply:
> @scottschurr may I ask why is Slice type used here, and not Buffer?

Sure.  The simple answer is that I was looking for the most obvious type to pass in.  The only call in the code base (currently) to `nft::changeTokenURI` looks like this:
```
    return nft::changeTokenURI(view(), owner, nftokenID, ctx_.tx[~sfURI]);
```
The awkward question is "What type does `ctx_.tx[~sfURI]` return?"  I could dig through a bunch of templates in `STObject` and try to figure that out.  But it's really easy to get lost in there.  So in situations like this I usually rely on my compiler to tell me.

I picked a type that I'm pretty sure is not constructible or assignable from whatever `ctx_.tx[~sfURI]` returns.  I don't remember the specific type I picked this time, but I often pick `int*`.  Then I tell the compiler that I want to assign `ctx_.tx[~sfURI]` to an `int*`.  It might look like this:
```
    [[maybe_unused]] int* junk = ctx_.tx[~sfURI];  // !!!! DEBUG !!!!
```
Then I compile hoping to get a useful error message out.  I use clang, so I this case I got:
```
/Users/scott/Development/rippled/src/xrpld/app/tx/detail/NFTokenModify.cpp:87:27: error: no viable conversion from 'std::optional<std::decay_t<typename STBlob::value_type>>' (aka 'optional<ripple::Slice>') to 'int *'
    [[maybe_unused]] int* junk = ctx_.tx[~sfURI];
```
Way off at the end of the error message it tells me "no viable conversion from ...  (aka `optional<ripple::Slice>`) to `int *`.

So the error message told me that, in order to avoid conversions, I needed to pass an `optional<ripple::Slice>` to `changeTokenURI()`.  Once I had that information I removed the test code from the code base.

A `Slice` is really cheap to copy, so it often makes sense to pass a `Slice` by value.  But I'm all the way up to an `optional<ripple::Slice>` which is not quite so obvious about pass-by-value or pass-by-const-ref.  I was conservative and picked pass-by-const-ref.

> Also another related question, in `STBlob`, it uses `Buffer` to store `value_`, but why `value_type` have type `Slice` and not `Buffer`?

That's a great question, and it's all about efficiency.  If we look inside `Buffer` it holds onto the buffer contents like this:
```
    std::unique_ptr<std::uint8_t[]> p_;
```
It's not a `shared_ptr`, it's a `unique_ptr`.  So the `Buffer` holds a `unique_ptr` to the only copy of the data.  If you're not acquainted with `unique_ptr`, it's worth getting to know.  Here's what CppReference has to say: https://en.cppreference.com/w/cpp/memory/unique_ptr

As CppReference notes, a `unique_ptr` is neither copy constructible or copy assignable.  So we _can't_ copy the `Buffer`.  The way it holds its data does not allow us to do so.  We have to get information about the contents of the `Buffer` some other way.

That way is called a `Slice`.  If you look inside a `Slice`, you'll see that it only holds two pieces of data:
```
private:
    std::uint8_t const* data_ = nullptr;
    std::size_t size_ = 0;
```
We have a pointer to the `const` data (so we can't mess with the data inside the `Buffer`) and a length.  If this code had been written with C++20, we'd probably be using the `ranges` library.  A `Slice` looks very much like a `ranges::contiguous_range`.  But this was done using C++11.

So `Slice` gives us a way to pass around information about the contents of the `Buffer` without actually copying `Buffer` or moving its contents around.

Why go to all this effort instead of putting the data inside of a `shared_ptr` and passing around `shared_ptr`s?  We gain both efficiency and safety.

**Efficiency**

A `shared_ptr` requires atomic operations, which are compute expensive.  If you copy a `shared_ptr` you have to increment its `atomic` count.  And every time a `shared_ptr` is destroyed its `atomic` count is decremented.  The `unique_ptr` has none of that overhead.

**Safety**
A `shared_ptr` is risky in two ways.

1. First is that if you have multiple `shared_ptr`s to non-`const` data, then anybody with one of those `shared_ptr` can modify the contents of the `Buffer`.  That leads to debugging mayhem if anyone doesn't follow rules closely.  That's why a `Slice` holds a `std::uint8_t const* data_`.  That `const` keeps anyone holding a`Slice` from messing with the contents of the `Buffer`.

2. The other problem with `shared_ptr` is that the contents of the `shared_ptr` hang around until the last copy of that `shared_ptr` is destroyed.  It is often difficult to know who has responsibility for destroying the contents of a `shared_ptr`.  But a `unique_ptr` has none of that ambiguity.

With `unique_ptr` the lifetime is obvious.  Whoever holds the `unique_ptr` has responsibility for the lifetime.  If the lifetime needs to exceed the lifetime of the holding object, then the `unique_ptr` must be _moved_, so the originally holding object no longer has the contents of the `unique_ptr`.

That's a very long winded answer.  I hope it made some sense.

Reply:
Really appreciate the detailed response. Thanks Scott! ❤️